### ISSUE TITLE
Sagemath 8.1

### ISSUE THREAD

**Issue by Hans-MartinHannibalLauridsen** [defeo/ss-isogeny-software#5]
It would seem that there is some issue with Cython, it can't seem to find the .pxd file when using cimport.

**Comment by Hans-MartinHannibalLauridsen** [comment]
It would seem that cocalc's online version of sagemath, lingers from the same issue, but I think that it is running sage 8.1 and python 2.7 too.

**Comment by defeo** [comment]
Hello, thanks for the report.

I guess this code has been broken at least since Sage 8.0 (it would be much easier if I could set up Travis, but I'm not sure this is feasible yet).

I could fix all bugs in  c510c53, except one. Sage seems to have changed the way it compiles `.spyx` files, so it does not find `gfp2.c` during Cython compilation anymore (this whole `.spyx` thing was a hack anyway, I'm surprised it worked for so long). I could fix it locally by manually setting the (absolute) path in `gfp2.pxd`:

```diff
diff --git a/gfp2.pxd b/gfp2.pxd
index 44a76b0..57105ed 100644
--- a/gfp2.pxd
+++ b/gfp2.pxd
@@ -13,7 +13,7 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-cdef extern from "gfp2.c":
+cdef extern from '/home/dfl/ss-isogeny-software/gfp2.c':
 
   ctypedef struct __mpz_struct:
     pass
```

Obviously this is not a fix I can push to the repo, but you can use it to test the code on your machine.

Do you happen to know how to fix this? Maybe @jdemeyer could be of help here?

**Comment by defeo** [comment]
May I ask you why you are interested in this code? https://github.com/Microsoft/PQCrypto-SIDH is much more *au goût du jour*.

**Comment by jdemeyer** [comment]
Upstream: https://trac.sagemath.org/ticket/24764

Quick fix:
```diff
diff --git a/src/sage/misc/cython.py b/src/sage/misc/cython.py
index 6a3b051..0fbcfd1 100644
--- a/src/sage/misc/cython.py
+++ b/src/sage/misc/cython.py
@@ -554,7 +554,7 @@ def cython(filename, verbose=0, compile_message=False,
         os.chdir(target_dir)
         ext, = cythonize([ext],
                          aliases=cython_aliases(),
-                         include_path=includes,
+                         include_path=[orig_cwd] + includes,
                          quiet=not verbose)
     except CompileError:
         # Check for names in old_pxi_names
```

Even after fixing that, your package fails to run because of
```
/home/jdemeyer/.sage/temp/tamiyo/4132/spyx/_tmp_ss_isogeny_software_pqcrypto11_spyx/_tmp_ss_isogeny_software_pqcrypto11_spyx_0.pyx in init _tmp_ss_isogeny_software_pqcrypto11_spyx_0()
     33 from sage.schemes.elliptic_curves.constructor import EllipticCurve
     34 from sage.categories.finite_fields import FiniteFields
---> 35 from sage.rings.arith import is_square
     36 from sage.misc.functional import is_odd
     37 

ImportError: No module named arith
```

**Comment by defeo** [comment]
Thanks @jdemeyer. (after fixing `is_square` and another couple of bugs) your upstream fix solves the problem.

@Hans-MartinHannibalLauridsen, it seems you will have to wait for the next SageMath beta, or apply the patch given above by @jdemeyer.

**Comment by Hans-MartinHannibalLauridsen** [comment]
Oh I see. Have you been collaborating with the contributors on the work done at:  https://github.com/Microsoft/PQCrypto-SIDH ?

Thanks alot! I'll have a look at it during this week.

I'm doing my master thesis with the headline _isogenies and their cryptographic application('s)_, basically revolving around your work on the matter.

**Comment by defeo** [comment]
> Oh I see. Have you been collaborating with the contributors on the work done at: https://github.com/Microsoft/PQCrypto-SIDH ?

I have nothing to do with the code, but, yes, we've submitted together to the NIST competition.

> I'm doing my master thesis with the headline isogenies and their cryptographic application('s), basically revolving around your work on the matter.

It is indeed true that this code is a bit more general than the usual SIDH implementations, such as Microsoft's, however be aware that it is very buggy, and in some respect outdated.

**Comment by fahmi100ye** [comment]
pqcrypto11.pyx:18:0: 'gfp2.pxd' not found

how can fix it

**Comment by Hans-MartinHannibalLauridsen** [comment]
The above fix did it for me. Just go to your sage folder and search for the cython.py file  and make the changes jdemeyer posted earlier in the thread. 

Well updating and restarting as usual might not hurt anybody. 

Best of luck.

**Comment by defeo** [comment]
Let me stress that the issue is solved in the next stable release of Sage (8.2), and the fix is already available in the release candidate.
